jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/FileMappedHttpContentFactory.java

```diff
@@ -0,0 +1,72 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FileMappedHttpContentFactory implements HttpContent.Factory
+{
+    private static final Logger LOG = LoggerFactory.getLogger(FileMappedHttpContentFactory.class);
+
+    private final HttpContent.Factory _factory;
+
+    public FileMappedHttpContentFactory(HttpContent.Factory factory)
+    {
+        _factory = Objects.requireNonNull(factory);
+    }
+
+    @Override
+    public HttpContent getContent(String path) throws IOException
+    {
+        HttpContent content = _factory.getContent(path);
+        if (content != null && content.getContentLengthValue() > 16 * 1024)
+            return new FileMappedContent(content);
+        return content;
+    }
+
+    public static class FileMappedContent extends HttpContentWrapper
+    {
+        private final HttpContent content;
+
+        public FileMappedContent(HttpContent content)
+        {
+            super(content);
+            this.content = content;
+        }
+
+        @Override
+        public RetainableByteBuffer getBuffer()
+        {
+            try
+            {
+                ByteBuffer buffer = BufferUtil.toMappedBuffer(content.getResource().getPath());
+                return new RetainableByteBuffer(buffer, b -> {});
+            }
+            catch (Throwable t)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Error getting Mapped Buffer", t);
+            }
+
+            return super.getBuffer();
```

Comment:
Here's some cases where BufferUtil.toMappedBuffer(Path) will not work.

* Extremely large files (I'm talking huge, like 200GB+)
* Real OS FileSystems based on dynamic content. (filesystems that have content that doesn't exist until it's read from, and can be different with each read, many such things exist on Windows)
* Emulated OS FileSystems. (eg: a Windows machine, in the WSL layer, reading a windows drive via the WSL layer)
* Some OS based Network FileSystems.
* Java FileSystems based on virtual concepts (eg:zipfs)

Reply:
Is there a test we can do to see if it is mappable? If so this should be done in `getContent()`.

Or I could try and call `BufferUtil.toMappedBuffer` in `getContent()` and only wrap if this does not throw. But if the file changes on disk it could become unmappable even after this `FileMappedContent` has been created.

If we cache a mapped buffer the size of the buffer could change if the file is changed on disk, so should we not cache any `FileMappedContent`.

Reply:
I think the easiest way, in a testcase, is to have a zipfs (mounted jar) and try to use BufferUtil.toMappedBuffer(Path), which should fail in 100% of use cases.